### ISSUE TITLE
[ClusterLoader2][NodeKiller] NodeKiller can leave broken nodes

### ISSUE THREAD

**Issue by jprzychodzen** [kubernetes/perf-tests#1130]
When a testing run uses NodeKiller, some nodes can be left in a broken state - with disabled services. This can influence results while running multiple testing jobs on the same cluster.

It happens when testing finishes before NodeKiller reboot is triggered.

Sample log:
```
...
I0323 08:03:00.196712   16174 resource_gather_worker.go:89] Closing worker for e2e-test-jprzychodzen-minion-group-2zfc
I0323 08:03:00.197101   16174 container_resource_gatherer.go:182] Waitgroup finished.
I0323 08:03:00.237245   16174 system_pod_metrics.go:124] collecting system pod metrics...
I0323 08:03:00.394737   16174 system_pod_metrics.go:228] Loaded restart count threshold overrides: map[fluentd-gcp:1 kube-proxy:1 metadata-p
roxy:1 prometheus-to-sd-exporter:1]
I0323 08:03:00.396502   16174 simple_test_executor.go:150] Step "Collecting measurements" ended
I0323 08:03:00.411423   16174 simple_test_executor.go:114] Summary of Chaos Monkey execution
NodeKiller: Recorded following events
NodeKiller: At Mon Mar 23 07:56:55 UTC 2020 stopService happend for node e2e-test-jprzychodzen-minion-group-49jb
I0323 08:03:00.411453   16174 probes.go:133] Stopping DnsLookupLatency probe...
I0323 08:03:20.715459   16174 probes.go:133] Stopping InClusterNetworkLatency probe...
I0323 08:03:46.222799   16174 simple_test_executor.go:370] Resources cleanup time: 45.811343198s
I0323 08:03:46.222842   16174 clusterloader.go:189] --------------------------------------------------------------------------------
I0323 08:03:46.222848   16174 clusterloader.go:190] Test Finished
I0323 08:03:46.222853   16174 clusterloader.go:191]   Test: testing/load/config.yaml
I0323 08:03:46.222859   16174 clusterloader.go:192]   Status: Success
I0323 08:03:46.222864   16174 clusterloader.go:196] --------------------------------------------------------------------------------
I0323 08:03:46.223239   16174 prometheus.go:181] Tearing down prometheus stack
```
In this sample NodeKiller did not schedule reboot an ClusterLoader2 has ended working, leaving e2e-test-jprzychodzen-minion-group-49jb without reboot.

NodeKiller should clean up disabled nodes by triggering reboot at the end of the test, regardless how long pending nodes should be disabled.

This should require changes in [clusterloader2/pkg/chaos](../tree/master/clusterloader2/pkg/chaos).

**Comment by mcastellin** [comment]
@jprzychodzen I would like to try have a go at it. It's my first contribution so I'll explain what I have in mind:
I've analyzed the issue a bit and I propose we listen for interruption on the `stopCh` in `nodes.go#kill` so we can break the downtime and have the Monkey schedule the restart. 
This would mean we also need to change the test executor to close the `stopCh` before printing the test summary. How does that sound?

**Comment by jprzychodzen** [comment]
Hi @mcastellin, thanks for expressing an interest to contribute to this issue.

Changing `nodes.go#kill` sounds great and you are right, there needs to be some modification to [simple_test_executor.go](../tree/master/clusterloader2/pkg/test/simple_test_executor.go). 

Closing `stopCh` will generate a rebooting of nodes in random order (between/after summary logs) and ideally we would like to prevent that. Could you also add some guard to make sure that no go routine is running inside [nodes.go#kill](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/chaos/nodes.go#L145) and make sure that summary is shown only then? Some kind of synchronization between `#kill` and `#Summary`.

**Comment by mcastellin** [comment]
@jprzychodzen thanks for you inputs, I'll do my best to implement what you're asking. 
Can I reach out to you on the Slack community to ask a few questions on how to test the changes before committing? 
I see the bot is scraping PRs to run against the GCE cluster, I wonder if you guys perform any test on your local machine first.

**Comment by jprzychodzen** [comment]
Few packages have unit tests implemented, however it seems that Node Killer is missing them. If you could add test provider to fake Node Killer actual node disablement and reboot so it could run locally, it would be a great improvement and would unblock local testing of a Node Killer.

When updating Node Killer I've mostly run a small cluster and test Node Killer changes with a synthetic ClusterLoader configuration - enabled Chaos Monkey as in [load config](../tree/master/clusterloader2/testing/load/config.yaml) with a Sleep, like in [access-tokens config](../tree/master/clusterloader2/testing/access-tokens/config.yaml).

I've joined [sig-scalability](https://kubernetes.slack.com/messages/sig-scalability) Slack, you should be able to ask some quick questions there, although I guess it would be better to keep as much discussion here. This way we will have some written archive for the future reference.

**Comment by jprzychodzen** [comment]
/closed
Resolved in #1135

**Comment by jprzychodzen** [comment]
Created a refactoring issue #1137 as a follow up.
